### PR TITLE
LTC 4 cards

### PR DESCRIPTION
**Pull request by Simisays**
final LTC cards that don't require engine changes. Though these need some more work

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Name:Gandalf, Westward Voyager
+ManaCost:3 G U
+Types:Legendary Creature Avatar Wizard
+PT:5/5
+T:Mode$ SpellCast | ValidCard$ Card.cmcGE5 | ValidActivatingPlayer$ You | Execute$ TrigReveal | TriggerZones$ Battlefield | TriggerDescription$ Whenever you cast a spell with mana value 5 or greater, each opponent reveals the top card of their library. If any of those cards shares a card type with that spell, copy that spell, you may choose new targets for the copy, and each opponent draws a card. Otherwise, you draw a card. (A copy of a permanent spell becomes a token.)
+SVar:TrigReveal:DB$ RepeatEach | RepeatPlayers$ Opponent | RepeatSubAbility$ DBReveal | SubAbility$ DBChoose
+SVar:DBReveal:DB$ PeekAndReveal | Defined$ Remembered | ImprintRevealed$ True
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt:7]: way too complicated + you get to cast your own spell, not the others...

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Name:Mirkwood Trapper
+ManaCost:1 G U
+Types:Creature Elf Scout
+PT:1/4
+T:Mode$ Attacks | ValidCard$ Creature | Attacked$ You | TriggerZones$ Battlefield | Execute$ TrigPump | TriggerDescription$ Whenever a player attacks you, target attacking creature gets -2/-0 until end of turn.
+SVar:TrigPump:DB$ Pump | ValidTgts$ Creature.ControlledBy TriggeredAttackingPlayer | NumAtt$ -2 | IsCurse$ True
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:6]: wrong Tgts

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Name:Mirkwood Trapper
+ManaCost:1 G U
+Types:Creature Elf Scout
+PT:1/4
+T:Mode$ Attacks | ValidCard$ Creature | Attacked$ You | TriggerZones$ Battlefield | Execute$ TrigPump | TriggerDescription$ Whenever a player attacks you, target attacking creature gets -2/-0 until end of turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:5]: wrong trigger

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+SVar:TrigPump:DB$ Pump | ValidTgts$ Creature.ControlledBy TriggeredAttackingPlayer | NumAtt$ -2 | IsCurse$ True
+T:Mode$ AttackersDeclaredOneTarget | ValidAttackers$ Creature.YouCtrl | AttackedTarget$ Player.Other | Execute$ TrigChoose | TriggerZones$ Battlefield | TriggerDescription$ Whenever a player attacks, if they aren't attacking you, that player chooses an attacking creature. It gets +2/+0 until end of turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:7]: wrong trigger

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+Name:Mirkwood Trapper
+ManaCost:1 G U
+Types:Creature Elf Scout
+PT:1/4
+T:Mode$ AttackersDeclared | ValidCard$ Creature | Attacked$ You | TriggerZones$ Battlefield | Execute$ TrigPump | TriggerDescription$ Whenever a player attacks you, target attacking creature gets -2/-0 until end of turn.
+SVar:TrigPump:DB$ Pump | ValidTgts$ Creature.attacking | NumAtt$ -2 | IsCurse$ True
+T:Mode$ AttackersDeclared | ValidAttackers$ Creature.YouCtrl | AttackedTarget$ Player.Other | Execute$ TrigChoose | TriggerZones$ Battlefield | TriggerDescription$ Whenever a player attacks, if they aren't attacking you, that player chooses an attacking creature. It gets +2/+0 until end of turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:7]: the Valid doesn't make sense
+ you need to write it as condition instead or an opponent could get both triggers at the same time

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+Name:Mirkwood Trapper
+ManaCost:1 G U
+Types:Creature Elf Scout
+PT:1/4
+T:Mode$ AttackersDeclared | ValidCard$ Creature | Attacked$ You | TriggerZones$ Battlefield | Execute$ TrigPump | TriggerDescription$ Whenever a player attacks you, target attacking creature gets -2/-0 until end of turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:5]: bad params

File: forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+SVar:TrigReveal:DB$ PeekAndReveal | NoPeek$ True | Defined$ Opponent | RevealValid$ Card.sharesCardTypeWith TriggeredCard | AnyNumber$ True | ImprintRevealed$ True | SubAbility$ DBCopyCast
+SVar:DBCopyCast:DB$ CopySpellAbility | Defined$ Imprinted | Amount$ 1 | MayChooseTarget$ True | Optional$ True | RememberNewCard$ True | SubAbility$ DBDrawAll
+SVar:DBDrawAll:DB$ Draw | Defined$ Opponent | ConditionDefined$ Remembered | ConditionPresent$ Card | ConditionCompare$ EQ1 | SubAbility$ DBDrawYou
+SVar:DBDrawYou:DB$ Draw | ConditionDefined$ Remembered | ConditionPresent$ Card | ConditionCompare$ EQ0 | SubAbility$ DBCleanup
+SVar:DBCleanup:DB$ Cleanup | ClearImprinted$ True | ClearRemembered$ True 
+DeckHas:Ability$Token
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt:11]: wrong Ability

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+ManaCost:1 G U
+Types:Creature Elf Scout
+PT:1/4
+T:Mode$ AttackersDeclared | AttackedTarget$ You | TriggerZones$ Battlefield | Execute$ TrigPump | TriggerDescription$ Whenever a player attacks you, target attacking creature gets -2/-0 until end of turn.
+SVar:TrigPump:DB$ Pump | ValidTgts$ Creature.attacking | NumAtt$ -2 | IsCurse$ True
+T:Mode$ AttackersDeclared | AttackedTarget$ Player.Other | Execute$ TrigChoose | NoResolvingCheck$ True | IsPresent$ Creature.attackingYou | PresentCompare$ EQ0 | TriggerZones$ Battlefield | TriggerDescription$ Whenever a player attacks, if they aren't attacking you, that player chooses an attacking creature. It gets +2/+0 until end of turn.
+SVar:TrigChoose:DB$ ChooseCard | Chooser$ TriggeredAttackingPlayer | Choices$ Creature.attacking | SubAbility$ DBPump 
+SVar:DBPump:DB$ Pump | Defined$ ChosenCard | NumAtt$ +2 | SubAbility$ DBCleanup
+SVar:DBCleanup:DB$ Cleanup | ClearChosenCard$ True
+AI:RemoveDeck:All
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:11]: what's the AI problem?

File: forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt (hunk 7 above)
Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt:11]: I was thinking about this (_A copy of a permanent spell becomes a token._)

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt:11]: hmn ok

File: forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt (hunk 1 above)
Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt]: you are still copying the wrong thing

Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt]: Oh you copy _your_ spell 🤦

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt]: the RevealValid still seems wrong

Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt]: you mean the param itself or the content of the param

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt]: all cards should be revealed
so probably just do it like the card says in the conditions

File: forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt [hunk 9]
```diff
@@ -0,0 +1,12 @@
+Name:Gandalf, Westward Voyager
+ManaCost:3 G U
+Types:Legendary Creature Avatar Wizard
+PT:5/5
+T:Mode$ SpellCast | ValidCard$ Card.cmcGE5 | ValidActivatingPlayer$ You | Execute$ TrigReveal | TriggerZones$ Battlefield | TriggerDescription$ Whenever you cast a spell with mana value 5 or greater, each opponent reveals the top card of their library. If any of those cards shares a card type with that spell, copy that spell, you may choose new targets for the copy, and each opponent draws a card. Otherwise, you draw a card. (A copy of a permanent spell becomes a token.)
+SVar:TrigReveal:DB$ PeekAndReveal | Defined$ Opponent | NoPeek$ True | ImprintRevealed$ True | SubAbility$ DBCopyCast
+SVar:DBCopyCast:DB$ CopySpellAbility | Defined$ TriggeredSpellAbility | ConditionDefined$ Imprinted | ConditionPresent$ Card.sharesCardTypeWith TriggeredCardLKICopy | MayChooseTarget$ True | Optional$ True | RememberNewCard$ True | SubAbility$ DBDrawAll
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/gandalf_westward_voyager.txt:7]: nothing says that it's Optional

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt [hunk 10]
```diff
@@ -0,0 +1,11 @@
+Name:Mirkwood Trapper
+ManaCost:1 G U
+Types:Creature Elf Scout
+PT:1/4
+T:Mode$ AttackersDeclared | AttackedTarget$ You | TriggerZones$ Battlefield | Execute$ TrigPump | TriggerDescription$ Whenever a player attacks you, target attacking creature gets -2/-0 until end of turn.
+SVar:TrigPump:DB$ Pump | ValidTgts$ Creature.attacking | NumAtt$ -2 | IsCurse$ True
+T:Mode$ AttackersDeclared | AttackedTarget$ Player.Other | Execute$ TrigChoose | NoResolvingCheck$ True | IsPresent$ Creature.attackingYou | PresentCompare$ EQ0 | TriggerZones$ Battlefield | TriggerDescription$ Whenever a player attacks, if they aren't attacking you, that player chooses an attacking creature. It gets +2/+0 until end of turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:7]: Still wrong:
> If a player is attacking a planeswalker you control or a battle you're protecting but not attacking you, Mirkwood Trapper's last ability will still trigger.

Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt:7]: Why NoResolvingCheck?

File: forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt [hunk 11]
```diff
@@ -0,0 +1,14 @@
+Name:Gilraen, Dunedain Protector
+ManaCost:2 W
+Types:Legendary Creature Human Noble
+PT:2/3
+A:AB$ ChangeZone | Cost$ 2 T | ValidTgts$ Creature.Other+YouCtrl | Origin$ Battlefield | Destination$ Exile | TgtPrompt$ Select another target creature you control | SubAbility$ DBReturn | SpellDescription$ Exile another target creature you control. You may return that card to the battlefield under its owner's control. If you don't, at the beginning of the next end step, return that card to the battlefield under its owner's control with a vigilance counter and a lifelink counter on it.
+SVar:DBReturn:DB$ ChangeZone | Defined$ Targeted | Origin$ All | Optional$ True | Destination$ Battlefield | SubAbility$ DBDelayedTrigger
+SVar:DBDelayedTrigger:DB$ DelayedTrigger | RememberObjects$ Targeted | Mode$ Phase | Phase$ End Of Turn | Execute$ TrigReturn | SpellDescription$ If you don't, at the beginning of the next end step, return that card to the battlefield under its owner's control with a vigilance counter and a lifelink counter on it.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt:7]: this looks like the delayed trigger would always fire, that's wrong

Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt:7]: can't figure this one out, only thing that might work is a storevar with rememberchanged on line 6 ?

File: forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt (hunk 10 above)
Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/mirkwood_trapper.txt]: It doesn't really make sense to be over specific, just remove AttackedTarget

File: forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt (hunk 11 above)
Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt]: You are remembering if it gets returned, then your condition checks if something got remembered:
that's the exact opposite of what the card wants

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt]: I'd say that most likely didn't work before either :/
Check _Semester's End_ for pointers

Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt]: finally 
image
🥳

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt]: Yes but your copy&paste still needs work:
- card isn't called Semester's End
- nothing says it should only get counters if it's still a creature

Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt]: thanks again, only thing that might still need work is that the cleanup doesn't happen at the right time ?
image

Comment (Simisays) [reply forge-gui/res/cardsfolder/upcoming/gilraen_dunedain_protector.txt]: fixed now

**Review by tool4ever — APPROVED** [review]

**Pull request merged**